### PR TITLE
Fix plotters

### PR DESCRIPTION
**Pull request by nikita-klsh**
- [x] predefine arguments (`cmap, vmax, vmin, aspect`) for plotters 
- [x] plotters `index` argument replaced by `pos` 
now it is the position of the gather in the batch, i.e. 0, 1, ..., len(batch), not the fair index(FFID). `pos=0` by default 
- [x] unify saving(a bit)
- [x] rm blank lines on the boarders of the plots
- [x]  label axis ( for y axis time/sample units, for x based on some trace attribute. or based on [tickers](https://matplotlib.org/api/ticker_api.html#matplotlib.ticker.FixedFormatter))
- [x] additional small attribute plot  on top of the gather plot
- [x] plot multiple gathers via single plot call

-----

image

----

image

---
Specify axis ticks labels via `x_ticker` and  `y_ticker`: x -offset, y - time
image

---

More flexible way to adjust axis - specify **where** and **what** would be displayed via _Locator_ and _Formatter_ 
image

---

On top of the seismogramm plot how some attribute vary from trace to trace via `src_attribute`
Plot the horizons or any events on the seismogram itself via `src_event`
image

---

There can be multiple events passed.
image

---

**Note**: in case multiple events and multiple sources of seismogramms passed, the events are splitter among seismograms.

image

### CODE REVIEW

File: seismicpro/src/plot_utils.py [hunk 1]
```diff
@@ -2,14 +2,59 @@
...
 import matplotlib.pyplot as plt
+from matplotlib import patches
+from matplotlib.ticker import ScalarFormatter, AutoLocator, IndexFormatter, LinearLocator, FixedFormatter
 from matplotlib import patches, colors as mcolors
+from mpl_toolkits.axes_grid1 import make_axes_locatable
 
 from .utils import measure_gain_amplitude
 
 
+def setup_imshow(ax, arr, **kwargs):
```
Comment (a-arefina) [inline seismicpro/src/plot_utils.py:13]: what about renaming to `custom_imshow`?

File: seismicpro/src/utils.py [hunk 2]
```diff
@@ -792,3 +793,41 @@ def transform_to_fixed_width_columns(path, path_save=None, n_spaces=8, max_len=(
...
+        y_ticker['formatter'] = IndexFormatter(yticks_labels)
+
+    elif y_ticker == 'samples': # default ticker will be used
+        pass 
+
+    elif isinstance(y_tick, dict): 
+        y_ticker = y_tick
+    
+    elif isinstance(x_tick, (list, tuple, np.ndarray)): # ticker will be inferred later in utils.setup_tickers
+        y_ticker = x_tick
```
Comment (a-arefina) [inline seismicpro/src/utils.py:831]: Did you mean this&

```suggestion
        y_ticker = y_tick
```

File: seismicpro/src/utils.py [hunk 3]
```diff
@@ -4,11 +4,12 @@
...
 import functools
 
+import segyio
 import numpy as np
 import pandas as pd
-from sklearn.linear_model import LinearRegression
 from scipy.signal import medfilt, hilbert
-import segyio
+from matplotlib.ticker import IndexFormatter
+from sklearn.linear_model import LinearRegression
```
Comment (a-arefina) [inline seismicpro/src/utils.py:12]: do you use it?

File: seismicpro/src/plot_utils.py [hunk 4]
```diff
@@ -2,14 +2,59 @@
...
+        'cmap': 'gray',
+        'vmin': np.quantile(arr, 0.1),
+        'vmax': np.quantile(arr, 0.9),
+        'aspect': 'auto',
+        'extent': (0, arr.shape[1], arr.shape[0], 0),
+    }
+
+    ax.imshow(arr, **{**defaults, **kwargs})
+
+def setup_tickers(ax, x_ticker, y_ticker):
```
Comment (a-arefina) [inline seismicpro/src/plot_utils.py:26]: what about moving this to utils or moving `infer_axis_tickers` here?

**Review by a-arefina — COMMENTED** [review]
Good work!
Plz carefully review possible `arrs` types and shapes

do you think it makes sense to extract a common method to plot seismograms in `spectrum_plot`, `gain_plot` and `statistics_plot`?

File: seismicpro/src/plot_utils.py [hunk 5]
```diff
@@ -87,42 +137,44 @@ def seismic_plot(arrs, wiggle=False, xlim=None, ylim=None, std=1, # pylint: disa
...
-
-        ax[0, i].set_aspect('auto')
+            ax.scatter(*pts, s=s, c=scatter_color)
 
     if title is not None:
         fig.suptitle(title)
+
     if save_to is not None:
-        plt.savefig(save_to, dpi=dpi, transparent=True)
+        plt.savefig(save_to, dpi=dpi)
```
Comment (anton-br) [inline seismicpro/src/plot_utils.py:172]: condiser to change all savefigs to
```suggestion
        plt.savefig(save_to, dpi=dpi, bbox_inches='tight', pad_inches=0.1)
```

File: seismicpro/src/utils.py [hunk 6]
```diff
@@ -792,3 +793,41 @@ def transform_to_fixed_width_columns(path, path_save=None, n_spaces=8, max_len=(
...
+        x_ticker['formatter'] = IndexFormatter(xticks_labels)
+
+    elif isinstance(x_tick, dict): # dict with config passed, it will be used directly
+        x_ticker = x_tick
+    elif isinstance(x_tick, (list, tuple, np.ndarray)): # ticker will be inferred later in utils.setup_tickers
+        x_ticker = x_tick
+
+    # process y-ticker
+    if y_tick == 'time': # infer ticker from meta
+        yticks_labels = batch.meta[src]['samples']
```
Comment (anton-br) [inline seismicpro/src/utils.py:821]: consider to cast `time` to int (or cast to int only the numbers like 1.0, 2.0 etc)

File: seismicpro/src/plot_utils.py [hunk 7]
```diff
@@ -51,9 +51,17 @@ def scatter_on_top(ax, attribute):
     top_subax.set_xticks([])
     top_subax.yaxis.tick_right()
 
+def scatter_within(ax, points, **scatter_within_kwargs):
+    if np.isscalar(points[0]):
```
Comment (a-arefina) [inline seismicpro/src/plot_utils.py:55]: use is_1d_array

File: seismicpro/src/plot_utils.py [hunk 8]
```diff
@@ -106,17 +114,47 @@ def seismic_plot(arrs, wiggle=False, xlim=None, ylim=None, std=1, # pylint: disa
 
     line_color = 'k' if line_color is None else line_color
 
-    if np.isscalar(arrs[0][0]): # arrs is a single seismogramm i.e. 2darray, wrap it with another array with dtype='O' 
-        empty = np.empty((1, 1), dtype='O')
-        empty[0, 0] = arrs
-        arrs = empty
+    def is_2d_array(array):
+        return np.isscalar(array[0][0])
```
Comment (a-arefina) [inline seismicpro/src/plot_utils.py:118]: check that array is not 1d before getting items recursively

File: seismicpro/src/plot_utils.py [hunk 9]
```diff
@@ -106,17 +114,47 @@ def seismic_plot(arrs, wiggle=False, xlim=None, ylim=None, std=1, # pylint: disa
...
+    
+    def is_1d_array(array):
+        return np.isscalar(array[0])
 
-    fig, ax = plt.subplots(*arrs.shape, figsize=figsize, squeeze=False)
-    for i, (arr, ax) in enumerate(zip(arrs.flatten(), ax.flatten())):
+    def infer_array(cond, arrs, transpose=False, broadcast_to=None, unite_points=False):
```
Comment (a-arefina) [inline seismicpro/src/plot_utils.py:123]: add docstring plz

File: seismicpro/src/plot_utils.py [hunk 10]
```diff
@@ -106,17 +114,47 @@ def seismic_plot(arrs, wiggle=False, xlim=None, ylim=None, std=1, # pylint: disa
...
+            blank[0, 0] = arrs
+        elif cond(arrs[0]):
+            blank = np.empty((1, len(arrs)), 'O')
+            for i, _ in enumerate(arrs):
+                blank[0, i] = arrs[i]
+        elif cond(arrs[0][0]):
+            blank = np.empty((len(arrs), len(arrs[0])), 'O')
+            for i, _ in enumerate(arrs):
+                for j, _ in enumerate(arrs[0]):
+                    blank[i, j] = arrs[i][j]
```
Comment (a-arefina) [inline seismicpro/src/plot_utils.py:138]: Is it fast enough? Can this be done using numpy's expand_dims/reshape/stack, etc?

File: seismicpro/src/plot_utils.py [hunk 11]
```diff
@@ -144,7 +182,10 @@ def seismic_plot(arrs, wiggle=False, xlim=None, ylim=None, std=1, # pylint: disa
                 setup_imshow(ax, arr.T, **kwargs)
         
             if attribute is not None:
-                scatter_on_top(ax, attribute.flatten()[i])
+                scatter_on_top(ax, att.flatten()[i])
+
+            if pts is not None:
+                scatter_within(ax, pts.flatten()[i])
```
Comment (a-arefina) [inline seismicpro/src/plot_utils.py]: What about renaming `attribute` to extra_points_top and `pts` to extra_points_over, or smth similar? Is plot position the only difference between these two arrays?

Even more crasy idea: having extra_points as list of arrays and extra_points_scatter_position as list of 'top'/'over' flags :smiling_imp:

File: seismicpro/src/plot_utils.py (hunk 10 above)
Comment (nikita-klsh) [reply seismicpro/src/plot_utils.py]: It feels like numpy can do it in a few lines. However, I did not find out how to

File: seismicpro/src/plot_utils.py (hunk 11 above)
Comment (nikita-klsh) [reply seismicpro/src/plot_utils.py]: Too crazy. 
I try to name this arguments with more seismic meaning rather than describing how these objects are being plotted. 
Should we reconsider this?

**Comment by alexandrkuvaev** [comment]
Covered in https://github.com/gazprom-neft/SeismicPro/pull/64
